File: main/java/gaiasbounty/world/WorldGenManager.java
package gaiasbounty.world;

import scala.collection.mutable.HashMap;

import java.util.Random;
import java.util.ArrayList;

import gaiasbounty.config.ConfigManager;
import gaiasbounty.world.gen.*;
import gaiasbounty.world.gen.IWorldGenBase;
import cpw.mods.fml.common.registry.GameRegistry;
import net.minecraft.world.World;
import net.minecraft.world.biome.BiomeGenBase;
import net.minecraft.world.chunk.IChunkProvider;
import net.minecraftforge.common.BiomeDictionary;
import net.minecraftforge.common.BiomeDictionary.Type;
import cpw.mods.fml.common.IWorldGenerator;

public class WorldGenManager implements IWorldGenerator
{
    public static final WorldGenManager INSTANCE = new WorldGenManager();

    private static IWorldGenBase[] generators = {new WorldGenApple(), new WorldGenPeach(), new WorldGenCherry(),
            new WorldGenBanana(), new WorldGenOrange(), new WorldGenLemon(), new WorldGenLime(), new WorldGenPlum(),
            new WorldGenMango(), new WorldGenPear(), new WorldGenPomegranate(), new WorldGenFig(),
            new WorldGenWalnut(), new WorldGenHazelnut(), new WorldGenPecan(), new WorldGenAlmond(),
            new WorldGenMacadamia(), new WorldGenMesquite(), new WorldGenOlive(), new WorldGenPalm(),
            new WorldGenSassafras(), new WorldGenTea(), new WorldGenRubber(), new WorldGenMaple()};
    private static IWorldGenBase[][] biomeGenerators;

    private WorldGenManager() {}

    public static void init()
    {
        GameRegistry.registerWorldGenerator(INSTANCE, 5);
    }

    public static void createBiomeAssociations()
    {
        BiomeGenBase[] biomes = BiomeGenBase.getBiomeGenArray();
        WorldGenManager.biomeGenerators = new IWorldGenBase[biomes.length][];
        for (int i = 0; i < biomes.length; i++)
        {
            if (biomes[i] != null)
            {
                ArrayList<IWorldGenBase> validGenerators = new ArrayList<IWorldGenBase>();
                for (int j = 0; j < generators.length; j++)
                {
                    if (generators[j].isSpawnConfigured() && generators[j].isValidSpawnBiome(biomes[i]))
                        validGenerators.add(generators[j]);
                }
                biomeGenerators[biomes[i].biomeID] = validGenerators.toArray(new IWorldGenBase[validGenerators.size()]);
            }
        }
    }

    public void generate(Random random, int chunkX, int chunkZ, World world,
                         IChunkProvider chunkGenerator, IChunkProvider chunkProvider)
    {
        IWorldGenBase[] biomeGenerators =
                WorldGenManager.biomeGenerators[world.getBiomeGenForCoords(chunkX * 16 + 8, chunkZ * 16 + 8).biomeID];
        for (int i = 0; i < biomeGenerators.length; i++)
        {
            if (biomeGenerators[i].canSpawnRandom(random))
                biomeGenerators[i].generate(random, chunkX, chunkZ, world, chunkGenerator, chunkProvider);
        }
    }
}

File: main/java/gaiasbounty/world/WorldGenManager.scala
package gaiasbounty.world;

import scala.collection.mutable.HashMap;
import java.util.Random;

import gaiasbounty.config.ConfigManager;
import gaiasbounty.world.gen._;
import gaiasbounty.world.IWorldGenHandler;
import cpw.mods.fml.common.registry.GameRegistry;
import net.minecraft.world.World;
import net.minecraft.world.biome.BiomeGenBase;
import net.minecraft.world.chunk.IChunkProvider;
import net.minecraft.world.gen.feature.WorldGenerator;
import net.minecraftforge.common.BiomeDictionary;
import net.minecraftforge.common.BiomeDictionary.Type;
import cpw.mods.fml.common.IWorldGenerator;

object WorldGenManager extends IWorldGenerator {
   private val generators = List(new WorldGenApple, new WorldGenPeach, new WorldGenCherry, new WorldGenBanana,
                         new WorldGenOrange, new WorldGenLemon, new WorldGenLime, new WorldGenPlum,
                         new WorldGenMango, new WorldGenPear, new WorldGenPomegranate, new WorldGenFig,
                         new WorldGenWalnut, new WorldGenHazelnut, new WorldGenPecan, new WorldGenAlmond,
                         new WorldGenMacadamia, new WorldGenMesquite, new WorldGenOlive, new WorldGenPalm,
                         new WorldGenSassafras, new WorldGenTea, new WorldGenRubber, new WorldGenMaple);
   private val biomes = new HashMap[Int, List[IWorldGenHandler]];
   
   def init() {
      GameRegistry registerWorldGenerator(this, 5);
   }
   
   def createBiomeAssociations() {
      BiomeGenBase.getBiomeGenArray().foreach(biome => {
         if (biome != null)
            biomes(biome.biomeID) = generators filter(generator => generator isValidSpawnBiome biome)
      });
   }
   
   def generate(random: Random, chunkX: Int, chunkZ: Int, world: World,
            chunkGenerator: IChunkProvider, chunkProvider: IChunkProvider) {
      biomes(world.getBiomeGenForCoords(chunkX * 16 + 8, chunkZ * 16 + 8).biomeID).foreach(generator => {
         if (generator canDoSpawn random)
            generator generate(random, chunkX, chunkZ, world, chunkGenerator, chunkProvider);
      });
   }
}
